File: app/src/main/java/com/example/client/screens/register/activity/RegisterActivity.java
package com.example.client.screens.register.activity;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.InputType;
import android.text.TextWatcher;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.ToggleButton;

import com.example.client.R;
import com.example.client.app.Constants;
import com.example.client.dialog.PrimaryDialog;
import com.example.client.dialog.VerificationDialog;
import com.example.client.screens.login.activity.LoginEmailActivity;
import com.example.client.screens.main.activity.MainActivity;
import com.example.client.screens.register.present.RegisterPresent;

public class RegisterActivity extends AppCompatActivity implements View.OnClickListener, IRegisterView {
    private ScrollView scrollView;
    private EditText fullname, phone, email, password, confirm;
    private ToggleButton eyePassword, eyeConfirm;
    private ImageView back;
    private TextView register;
    private TextView fullname_error, error_phone, error_email, error_password, error_confirm;
    private RegisterPresent rPresent;
    private ProgressBar progressBar;
    private PrimaryDialog dialog;
    private VerificationDialog verificationDialog;
    private RelativeLayout rllLoading;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);
        fullname = findViewById(R.id.fullname);
        phone = findViewById(R.id.phone);
        email = findViewById(R.id.email);
        password = findViewById(R.id.password);
        confirm = findViewById(R.id.confirm);
        eyePassword = findViewById(R.id.eyePassword);
        eyeConfirm = findViewById(R.id.eyeConfirm);
        back = findViewById(R.id.back);
        register = findViewById(R.id.register);
        fullname_error = findViewById(R.id.fullname_error);
        error_phone = findViewById(R.id.error_phone);
        error_email = findViewById(R.id.error_email);
        error_password = findViewById(R.id.error_password);
        error_confirm = findViewById(R.id.error_confirm);
        scrollView = findViewById(R.id.scrollView);
        progressBar = findViewById(R.id.progress_bar);
        rllLoading = findViewById(R.id.rll_loading);

        rPresent = new RegisterPresent(this);
        dialog = new PrimaryDialog();
        dialog.getInstance(this);
        verificationDialog = new VerificationDialog();
        verificationDialog.getInstance(this);
        password.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
        confirm.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
        eyePassword.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                password.setInputType(InputType.TYPE_CLASS_TEXT);
                password.setSelection(password.getText().length());
            } else {
                password.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
                password.setSelection(password.getText().length());
            }
        });

        eyeConfirm.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                confirm.setInputType(InputType.TYPE_CLASS_TEXT);
                confirm.setSelection(confirm.getText().length());
            } else {
                confirm.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
                confirm.setSelection(confirm.getText().length());
            }
        });

        fullname.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (s.length() < 6) {
                    fullname_error.setText("*Email < 6 ký tự");
                    fullname_error.setVisibility(View.VISIBLE);
                } else {
                    fullname_error.setText("");
                    fullname_error.setVisibility(View.GONE);
                }
                checkButton();
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        phone.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                String regex = "^(03|05|07|08|09|01[2|6|8|9])+([0-9]{8})\\b";
                if (phone.getText().toString().matches(regex)) {
                    error_phone.setText("");
                    error_phone.setVisibility(View.GONE);
                } else {
                    error_phone.setText("*Số điện thoại chưa đúng");
                    error_phone.setVisibility(View.VISIBLE);
                }
                checkButton();
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        email.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (s.length() < 6) {
                    error_email.setText("*Email < 6 ký tự");
                    error_email.setVisibility(View.VISIBLE);
                } else {
                    String regex = "^[\\w-\\.]+@([\\w-]+\\.)+[\\w-]{2,4}$";
                    if (email.getText().toString().matches(regex)) {
                        error_email.setText("");
                        error_email.setVisibility(View.GONE);
                    } else {
                        error_email.setText("*Email chưa đúng định dạng");
                        error_email.setVisibility(View.VISIBLE);
                    }
                }
                checkButton();
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        password.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (s.length() < 6) {
                    error_password.setText("*Password < 6 ký tự");
                    error_password.setVisibility(View.VISIBLE);
                } else {
                    error_password.setText("");
                    error_password.setVisibility(View.GONE);
                }
                checkButton();
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        confirm.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (!password.getText().toString().equals(confirm.getText().toString())) {
                    error_confirm.setText(R.string.password_invalid);
                    error_confirm.setVisibility(View.VISIBLE);
                    checkButton();
                    scrollView.post(() -> scrollView.scrollTo(0, scrollView.getHeight()));
                } else {
                    error_confirm.setText("");
                    error_confirm.setVisibility(View.GONE);
                    checkButton();
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        register.setOnClickListener(this);
        back.setOnClickListener(this);

    }

    private void checkButton() {
        if (fullname_error.getText().toString().length() == 0 &&
                error_phone.getText().toString().length() == 0 &&
                error_email.getText().toString().length() == 0 &&
                error_password.getText().toString().length() == 0 &&
                error_confirm.getText().toString().length() == 0) {
            setButton(true, R.drawable.bg_btn);
        } else {
            setButton(false, R.drawable.bg_btn_disable);
        }
    }

    private void setButton(boolean enable, int background) {

        register.setEnabled(enable);
        register.setBackgroundResource(background);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.back:
                finish();
                onBackPressed();
                break;
            case R.id.register:
                InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
                if (getCurrentFocus() != null) {
                    imm.hideSoftInputFromWindow(getCurrentFocus().getWindowToken(), 0);
                }
                rPresent.sendEmail(email.getText().toString());
                break;
        }
    }

    @Override
    public void register() {
        dialog.setDescription("Đăng ký thành công");
        dialog.hideBtnCancel();
        dialog.show();
        dialog.setOKListener(() -> {
            finish();
            LoginEmailActivity.loginEmailActivity.finish();
            startActivity(new Intent(this, MainActivity.class));
        });
    }

    @Override
    public void showVerificationDialog(boolean showDescription) {
        if (!showDescription) {
            verificationDialog.hideDescription();
        } else {
            verificationDialog.setDescription(getString(R.string.err_code_1010));
            verificationDialog.showDescription();
        }
        verificationDialog.setOKListener(() ->
                rPresent.verification(email.getText().toString(),
                        VerificationDialog.verificationCode,
                        fullname.getText().toString(),
                        phone.getText().toString(),
                        password.getText().toString()));
        verificationDialog.setCancelListener(() -> {
        });
        verificationDialog.show();
    }

    @Override
    public void showLoading() {
        rllLoading.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoading() {
        rllLoading.setVisibility(View.GONE);
    }

    @Override
    public void showErrorMessage(int errMessage) {
        dialog.setDescription(getString(errMessage));
        dialog.setOKListener(() -> {
        });
        dialog.hideBtnCancel();
        dialog.show();
    }
}

File: app/src/main/java/com/example/client/screens/register/activity/RegisterActivity.kt
package com.example.client.screens.register.activity

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.InputType
import android.text.TextWatcher
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.CompoundButton
import com.example.client.R
import com.example.client.base.BaseActivityMVP
import com.example.client.dialog.PrimaryDialog
import com.example.client.dialog.VerificationDialog
import com.example.client.screens.main.activity.MainActivity
import com.example.client.screens.register.present.IRegisterPresent
import com.example.client.screens.register.present.RegisterPresent
import kotlinx.android.synthetic.main.activity_register.*

class RegisterActivity : BaseActivityMVP<IRegisterPresent>(), View.OnClickListener, IRegisterView {

    override val presenter: IRegisterPresent
        get() = RegisterPresent(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)
    }

    override fun bindComponent() {
        et_password.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
        et_confirm.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
    }

    override fun bindEvent() {
        tbt_eye_password.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
            et_password.run {
                if (isChecked) {
                    inputType = InputType.TYPE_CLASS_TEXT
                    setSelection(text.length)
                } else {
                    inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
                    setSelection(text.length)
                }
            }
        }
        tbt_eye_confirm.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
            et_confirm.run {
                if (isChecked) {
                    inputType = InputType.TYPE_CLASS_TEXT
                    setSelection(text.length)
                } else {
                    inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
                    setSelection(text.length)
                }
            }
        }
        et_full_name.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                tv_full_name_error.run {
                    if (s.length < 6) {
                        text = context.getString(R.string.register_name_invalid)
                        visibility = View.VISIBLE
                    } else {
                        text = ""
                        visibility = View.GONE
                    }
                }
                checkButton()
            }

            override fun afterTextChanged(s: Editable) {}
        })
        et_phone.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val regex = "^(03|05|07|08|09|01[2|6|8|9])+([0-9]{8})\\b".toRegex()
                tv_phone_error.run {
                    if (et_phone.text.toString().matches(regex)) {
                        text = ""
                        visibility = View.GONE
                    } else {
                        text = context.getString(R.string.register_phone_invalid)
                        visibility = View.VISIBLE
                    }
                }
                checkButton()
            }

            override fun afterTextChanged(s: Editable) {}
        })
        et_email.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                tv_email_error.run {
                    if (s.length < 6) {
                        text = context.getString(R.string.register_email_length_invalid)
                        visibility = View.VISIBLE
                    } else {
                        val regex = "^[\\w-.]+@([\\w-]+\\.)+[\\w-]{2,4}$".toRegex()
                        if (et_email.text.toString().matches(regex)) {
                            text = ""
                            visibility = View.GONE
                        } else {
                            text = context.getString(R.string.register_email_invalid)
                            visibility = View.VISIBLE
                        }
                    }
                }
                checkButton()
            }

            override fun afterTextChanged(s: Editable) {}
        })
        et_password.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                tv_password_error.run {
                    if (s.length < 6) {
                        text = context.getString(R.string.register_password_length_invalid)
                        visibility = View.VISIBLE
                    } else {
                        text = ""
                        visibility = View.GONE
                    }
                }
                checkButton()
            }

            override fun afterTextChanged(s: Editable) {}
        })
        et_confirm.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                tv_confirm_error.run {
                    if (et_password.text.toString() != et_confirm.text.toString()) {
                        setText(R.string.register_confirm_password_invalid)
                        visibility = View.VISIBLE
                        scroll_view.post { scroll_view.scrollTo(0, scroll_view.height) }
                    } else {
                        text = ""
                        visibility = View.GONE
                    }
                }
                checkButton()
            }

            override fun afterTextChanged(s: Editable) {}
        })
        tv_register.setOnClickListener(this)
        imv_back.setOnClickListener(this)
    }

    private fun checkButton() {
        if ((tv_full_name_error.text.toString().isEmpty() &&
                        tv_phone_error.text.toString().isEmpty() &&
                        tv_email_error.text.toString().isEmpty() &&
                        tv_password_error.text.toString().isEmpty() &&
                        tv_confirm_error.text.toString().isEmpty()) ||
                (et_full_name.text.toString().isNotEmpty() &&
                        et_phone.text.toString().isNotEmpty() &&
                        et_email.text.toString().isNotEmpty() &&
                        et_password.text.toString().isNotEmpty() &&
                        et_confirm.text.toString().isNotEmpty())) {
            setButton(true, R.drawable.bg_btn)
        } else {
            setButton(false, R.drawable.bg_btn_disable)
        }
    }

    private fun setButton(enable: Boolean, background: Int) {
        tv_register.isEnabled = enable
        tv_register.setBackgroundResource(background)
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.imv_back -> {
                finish()
                onBackPressed()
            }
            R.id.tv_register -> {
                val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
                currentFocus?.let {
                    imm.hideSoftInputFromWindow(it.windowToken, 0)
                }
                presenter.sendEmail(et_email.text.toString(), et_phone.text.toString())
            }
        }
    }

    override fun register() {
        PrimaryDialog({
            finish()
            startActivity(Intent(this, MainActivity::class.java))
        }, {})
                .setDescription(getString(R.string.register_success))
                .show(supportFragmentManager)
    }

    override fun showVerificationDialog(isError: Boolean) {
        VerificationDialog({
            VerificationDialog.verificationCode?.let {
                presenter.verification(et_email.text.toString(), it,
                        et_full_name.text.toString(),
                        et_phone.text.toString(),
                        et_password.text.toString())
            }
        }, {})
                .setDescription(if (isError) getString(R.string.err_code_1010) else getString(R.string.dialog_verification_description))
                .show(supportFragmentManager)
    }

    override fun showLoading() {
        rll_loading.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        rll_loading.visibility = View.GONE
    }

    override fun onBackPress() {}

    override fun showErrorMessage(errMessage: Int) {
        showDialogErrorMessage(getString(errMessage))
    }

}